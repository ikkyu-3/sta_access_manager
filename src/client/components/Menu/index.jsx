// @flow
import type { MenuProps, MenuButton } from "@types";

import * as React from "react";
import {
  TOP,
  TOP_PATH,
  REGISTER,
  REGISTER_PATH,
  PARTICIPANT,
  PARTICIPANT_PATH
} from "@constants";

import styles from "./style.css";

const menuList: Array<MenuButton> = [
  { text: "入退室処理", name: TOP, path: TOP_PATH },
  { text: "ユーザ登録", name: REGISTER, path: REGISTER_PATH },
  { text: "利用者一覧", name: PARTICIPANT, path: PARTICIPANT_PATH }
];

function Menu(props: MenuProps): React.Node {
  const { selectedMenu, onButtonClick } = props;
  const { menu, list, button, selected } = styles;

  const listItems: Array<React.Node> = menuList.map(
    (item: MenuButton): React.Node => {
      const { name, path, text } = item;
      const className: string =
        name === selectedMenu ? `${button} ${selected}` : button;

      return (
        <li key={name}>
          <button
            className={className}
            onClick={() => {
              onButtonClick(path);
            }}
          >
            {text}
          </button>
        </li>
      );
    }
  );

  return (
    <nav className={menu}>
      <ul className={list}>{listItems}</ul>
    </nav>
  );
}
export default Menu;
