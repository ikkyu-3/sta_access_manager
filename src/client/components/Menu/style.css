@charset "utf-8";
.menu {
  position: absolute;
  width: 170px;
  height: 100%;
  top: 0;
  left: 0;
  background-color: whitesmoke;
  box-shadow: 1px 0px 4px 0px silver;
  z-index: 10;
}

.list {
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.button {
  width: 100%;
  height: 85px;
  margin: 0;
  padding: 0;
  font-size: 18px;
  color: gray;
  background: transparent;
  border: none;
  border-bottom: 1px dashed silver;
  box-sizing: border-box;
  transition: font-size 200ms;
  outline: none;
}

.button.selected {
  background-color: silver;
  border: none;
  color: white;
  font-size: 24px;
  font-weight: bold;
}
