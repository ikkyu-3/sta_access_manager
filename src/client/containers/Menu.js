// @flow
import type { State, Dispatch } from "@types";

import { connect } from "react-redux";
import { push } from "react-router-redux";

import Menu from "@components/Menu";

const mapStateToProps = (state: State) => ({
  selectedMenu: state.selectedMenu
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onButtonClick: (path: string) => {
    dispatch(push(path));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Menu);
