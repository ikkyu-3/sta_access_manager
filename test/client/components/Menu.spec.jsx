import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Menu from "@components/Menu";

const fn = () => {};

describe("Menu.jsxのテスト", () => {
  describe("スナップショット", () => {
    it("正しいレンダリング", () => {
      const tree = renderer
        .create(<Menu selectedMenu="top" onButtonClick={fn} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("コンポーネントのテスト", () => {
    it("currentに「入退室処理」を指定した時、「入退室処理」ボタンが選択されている", () => {
      const buttons = shallow(
        <Menu selectedMenu="top" onButtonClick={fn} />
      ).find("button");
      expect(buttons.at(0).hasClass("selected")).toBeTruthy();
      expect(buttons.at(1).hasClass("selected")).toBeFalsy();
      expect(buttons.at(2).hasClass("selected")).toBeFalsy();
    });

    it("currentに「ユーザ登録」を指定した時、「ユーザ登録」ボタンが選択されている", () => {
      const buttons = shallow(
        <Menu selectedMenu="register" onButtonClick={fn} />
      ).find("button");
      expect(buttons.at(0).hasClass("selected")).toBeFalsy();
      expect(buttons.at(1).hasClass("selected")).toBeTruthy();
      expect(buttons.at(2).hasClass("selected")).toBeFalsy();
    });

    it("currentに「利用者一覧」を指定した時、「利用者一覧」ボタンが選択されている", () => {
      const buttons = shallow(
        <Menu selectedMenu="participant" onButtonClick={fn} />
      ).find("button");
      expect(buttons.at(0).hasClass("selected")).toBeFalsy();
      expect(buttons.at(1).hasClass("selected")).toBeFalsy();
      expect(buttons.at(2).hasClass("selected")).toBeTruthy();
    });

    it("currentに何も指定しない時、どのボタンも選択されていない", () => {
      const buttons = shallow(<Menu selectedMenu="" onButtonClick={fn} />).find(
        "button"
      );
      expect(buttons.at(0).hasClass("selected")).toBeFalsy();
      expect(buttons.at(1).hasClass("selected")).toBeFalsy();
      expect(buttons.at(2).hasClass("selected")).toBeFalsy();
    });

    describe("buttonをクリック時", () => {
      const onButtonClick = jest.fn();
      let menu;

      beforeEach(() => {
        menu = shallow(<Menu selectedMenu="" onButtonClick={onButtonClick} />);
        onButtonClick.mockReset();
      });

      it("「入退室処理」をクリックした時、引数に'入退室処理'が渡される", () => {
        menu
          .find("button")
          .at(0)
          .simulate("click");
        expect(onButtonClick.mock.calls[0][0]).toBe("/");
      });

      it("「ユーザ登録」をクリックした時、引数に'ユーザ登録'が渡される", () => {
        menu
          .find("button")
          .at(1)
          .simulate("click");
        expect(onButtonClick.mock.calls[0][0]).toBe("/register");
      });

      it("「利用者一覧」をクリックした時、引数に'利用者一覧'が渡される", () => {
        menu
          .find("button")
          .at(2)
          .simulate("click");
        expect(onButtonClick.mock.calls[0][0]).toBe("/participant");
      });
    });
  });
});
