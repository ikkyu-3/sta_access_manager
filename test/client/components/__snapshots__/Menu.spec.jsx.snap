// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu.jsxのテスト スナップショット 正しいレンダリング 1`] = `
<nav
  className="menu"
>
  <ul
    className="list"
  >
    <li>
      <button
        className="button selected"
        onClick={[Function]}
      >
        入退室処理
      </button>
    </li>
    <li>
      <button
        className="button"
        onClick={[Function]}
      >
        ユーザ登録
      </button>
    </li>
    <li>
      <button
        className="button"
        onClick={[Function]}
      >
        利用者一覧
      </button>
    </li>
  </ul>
</nav>
`;
